mitosheet/mitosheet/step_performers/replace.py

```diff
@@ -40,48 +40,60 @@ def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Opti
 
-        if (any(df.dtypes == 'timedelta') and pd.__version__ < 1.4):
+        # Raise an error if the pandas version is too old to use timedelta with replace.
+        if (any(selected_columns.dtypes == 'timedelta') and pd.__version__ < 1.4):
             raise MitoError(
                 'version_error',
                 'Pandas version error',
                 'This version of pandas doesn\'t support replacing values in timedelta columns. Please upgrade to pandas 1.2 or later.',
             )
         
         try:
+            search_value_regex = f'(?i){search_value}'
             # Special case for boolean columns because when we convert to string, the values are 
             # all converted back to bool as True - even if the value is "False". 
-            if any(df.dtypes == 'bool'):
-                bool_columns = df.select_dtypes(include='bool')
-                non_bool_columns = df.select_dtypes(exclude='bool')
-                df[bool_columns.columns] = bool_columns.astype(str).replace(f'(?i){search_value}', replace_value, regex=True).applymap(cast_string_to_bool).astype(bool)
-                df[non_bool_columns.columns] = non_bool_columns.astype(str).replace(f'(?i){search_value}', replace_value, regex=True).astype(non_bool_columns.dtypes.to_dict())
+            if any(selected_columns.dtypes == 'bool'):
+                bool_columns = selected_columns.select_dtypes(include='bool')
+                non_bool_columns = selected_columns.select_dtypes(exclude='bool')
+                selected_columns[bool_columns.columns] = bool_columns.astype(str).replace(search_value_regex, replace_value, regex=True).applymap(cast_string_to_bool).astype(bool)
+                selected_columns[non_bool_columns.columns] = non_bool_columns.astype(str).replace(search_value_regex, replace_value, regex=True).astype(non_bool_columns.dtypes.to_dict())
             else:
-                df = df.astype(str).replace(f'(?i){search_value}', replace_value, regex=True).astype(df.dtypes.to_dict())
+                selected_columns = selected_columns.astype(str).replace(search_value_regex, replace_value, regex=True).astype(selected_columns.dtypes.to_dict())
 
             # Then, we replace the column headers in the state column_ids object
-            new_columns = df.columns.str.replace(f'(?i){search_value}', replace_value, regex=True)
-
+            # We convert the column headers to strings because the column headers can be any type
+            # Then we call astype because we need to convert them back to their original type. 
+            new_columns = pd.Index([re.sub(search_value_regex, replace_value, str(column)) for column in selected_columns.columns]).astype(selected_columns.columns.dtype).to_list()
             # Update the column headers in the state column_ids object
-            for old_column_name, new_column_name in zip(df.columns, new_columns):
+            for old_column_name, new_column_name in zip(selected_columns.columns, new_columns):
```

Comment:
selected_columns.columns illustrates that this isn't a great variable name choice here either.